core/server/common/src/main/java/alluxio/master/AbstractMaster.java

```diff
@@ -91,112 +95,78 @@ protected AbstractMaster(Journal journal, Clock clock,
-      // log should not exist, and all the logs will be complete.
-      mJournalWriter.completeLogs();
-
-      // Phase 3: Replay all the state of the checkpoint and the completed logs.
-      JournalTailer catchupTailer;
-      if (mStandbyJournalTailer != null && mStandbyJournalTailer.getLatestJournalTailer() != null
-          && mStandbyJournalTailer.getLatestJournalTailer().isValid()) {
-        // This master was previously in standby mode, and processed some of the journal. Re-use the
-        // same tailer (still valid) to continue processing any remaining journal entries.
-        LOG.info("{}: finish processing remaining journal entries (standby -> master).",
-            getName());
-        catchupTailer = mStandbyJournalTailer.getLatestJournalTailer();
-        catchupTailer.processNextJournalLogs();
-      } else {
-        // This master has not successfully processed any of the journal, so create a fresh tailer
-        // to process the entire journal.
-        catchupTailer = JournalTailer.Factory.create(this, mJournal);
-        if (catchupTailer.checkpointExists()) {
-          LOG.info("{}: process entire journal before becoming leader master.", getName());
-          catchupTailer.processJournalCheckpoint(true);
-          catchupTailer.processNextJournalLogs();
-        } else {
-          LOG.info("{}: journal checkpoint does not exist, nothing to process.", getName());
+      LOG.info("{}: Starting primary master.", getName());
+
+      // Step 1. Replay the journal entries.
+      long nextSequenceNumber =
+          mJournalCheckpointThread != null ? mJournalCheckpointThread.getNextSequenceNumber() : 0;
+      try (JournalReader journalReader = mJournal.getReader(
+          JournalReaderOptions.defaults().setPrimary(true)
+              .setNextSequenceNumber(nextSequenceNumber))) {
+        JournalEntry entry;
+        while ((entry = journalReader.read()) != null) {
+          processJournalEntry(entry);
         }
+        nextSequenceNumber = journalReader.getNextSequenceNumber();
+      } catch (InvalidJournalEntryException e) {
+        LOG.error("Invalid journal entry is found.", e);
+        // We found invalid journal, nothing we can do but crash.
+        throw new RuntimeException(e);
       }
-      long latestSequenceNumber = catchupTailer.getLatestSequenceNumber();
 
-      // Phase 4: initialize the journal and write out the checkpoint (the state of all
-      // completed logs).
-      try (JournalOutputStream checkpointStream =
-          mJournalWriter.getCheckpointOutputStream(latestSequenceNumber)) {
-        LOG.info("{}: start writing checkpoint.", getName());
-        streamToJournalCheckpoint(checkpointStream);
+      // Step 2: Start the journal writer.
+      mJournalWriter = mJournal.getWriter(
+          JournalWriterOptions.defaults().setNextSequenceNumber(nextSequenceNumber)
+              .setPrimary(true));
+      if (nextSequenceNumber == 0) {
```

Comment:
Please comment on why there is this special case.

Also, it seems like "step 2" is more involved than just starting a journal writer?